### PR TITLE
Implement barebones protocol config module

### LINKED ISSUE #164: Implement protocol config module logic
## Description

Enhance the `OpenLIFUProtocolConfig` module by implementing the following:

- Map widget buttons to new logic functions for protocol management.
- Add logic functions to get/set protocol values.
- Connect UI widgets to logic functions, including:
  - Populate a dropdown menu in the protocol configuration module with currently loaded protocols (from the `OpenLIFUData` module’s parameter node), including a "create new" option.
  - Implement save protocol functionality with overwrite dialog support.
  - Add delete protocol functionality.
  - Dynamically adjust widget availability based on values of certain widgets (e.g., `FocalPattern` determines available focal pattern parameters).

### References

- Depends on #163 and https://github.com/OpenwaterHealth/OpenLIFU-python/issues/196

### PR DESCRIPTION
**Pull request by arhowe00**
Closes #164

### CODE REVIEW

**Comment by ebrahimebrahim** [comment]
@arhowe00 Make sure to rebase this to the latest main, then ping us to review again. Currently this contains commits from #166

**Review by sadhana-r — CHANGES_REQUESTED** [review]
Great work! The saving and deleting of protocols to the database works well. I have made a few comments regarding style and organization of the code. And the way this module keeps track of the database connection needs to be updated. I'll do one more round of reviews once you have addressed the current comments.

**Comment by sadhana-r** [comment]
Another thought I had, currently `trigger_unsaved_changes` is triggered anytime a user edits the textboxes. Therefore, for example, if a user changes a value from 1 to 2 and then back to 1, it still shows as having 'unsaved changes'. This is probably a separate issue, but ideally, if a user is modifying an existing protocol, it would be nice to keep track of the original protocol and only trigger unsaved changes if whatever is entered is different from the original protocol. This would also allow us to later on add 'undo' or 'restore' options if helpful.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 1]
```diff
@@ -154,6 +225,187 @@ def onSceneEndClose(self, caller, event) -> None:
...
+            'Load protocol', # title of dialog
+            qsettings.value('OpenLIFU/databaseDirectory','.'), # starting dir, with default of '.'
+            "Protocols (*.json);;All Files (*)", # file type filter
+        )
+        if filepath:
+            protocol = openlifu_lz().Protocol.from_file(filepath)
+            self.logic.load_protocol_from_openlifu(protocol)  # load to memory
+            self.updateProtocolDisplayFromProtocol(protocol)  # update widget
+            self.ui.protocolSelector.setCurrentText(f"{protocol.name} (ID: {protocol.id})")  # select the protocol
+            self.updateWidgetSaveState(SaveState.NO_CHANGES)  # update save state
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:378]: Good points. I didn't implement the `Load Protocol` button exactly as it is done in the Data Module's logic.

To explain the additions, I assumed that if a user was in the protocol config module and hits "Load Protocol From File," in addition to loading the protocol in memory, I would also populate the widget fields, select the widget, and edit the save state.  You are also right that the save state edit is redundant because it is in `onDataParameterNodeModified`, which I can remove.  Let me know what you think about the user intuitiveness/interaction aspect and whether I should still not change the active (i.e. displayed and selected) protocol when a user loads protocol within the protocol config module.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 2]
```diff
@@ -154,6 +225,187 @@ def onSceneEndClose(self, caller, event) -> None:
...
 
+    def onDataParameterNodeModified(self, caller = None, event = None):
+
+        # If there is no database loaded, widget should be grayed out
+
+        if not self.logic.database_is_loaded():
+            self.setAllWidgetsEnabled(False)
+            self.ui.saveStateLabel.setProperty("text", "You must load a database to configure protocols.")
+            self.ui.saveStateLabel.setProperty("styleSheet", "color: blue; font-weight: bold; font-size: 16px; border: 3px solid blue; padding: 30px;")
+            return
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:236]: > This functionality doesn't work...

Yes, I noticed this while I was making commits on #169 .  I fixed it in [af89869](https://github.com/OpenwaterHealth/SlicerOpenLIFU/commit/af89869f299ef13ceb0f57d527cd87db41c95c25).  The reason I didn't add it in here is because in making changes to implement #169 .  What should I do in this case, add the commits here and then drop it from that branch?

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 3]
```diff
@@ -97,22 +150,40 @@ def setup(self) -> None:
...
+
+        # Connect main widget functions
+
+        self.ui.protocolSelector.textActivated.connect(self.onProtocolSelected)
+        self.ui.loadProtocolButton.clicked.connect(self.onLoadProtocolClicked)
+        self.ui.createNewProtocolButton.clicked.connect(self.onNewProtocolClicked)
+
+        self.ui.protocolSaveButton.clicked.connect(self.onSaveProtocolClicked)
+        self.ui.protocolDeleteButton.clicked.connect(self.onDeleteProtocolClicked)
+
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:185]: Yes, I will implement this change in #171 , I recently renamed the button as part of this PR

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:236]: If you need to fix something to make the branch of this PR work, then do that fix here (you can cherry pick it over). Then rebase the things you branched off of here to get that fix into your other branches. So _yes_ to your question if that is what you meant :)

**Comment by arhowe00** [comment]
> Another thought I had, currently `trigger_unsaved_changes` is triggered anytime a user edits the textboxes. Therefore, for example, if a user changes a value from 1 to 2 and then back to 1, it still shows as having 'unsaved changes'. This is probably a separate issue, but ideally, if a user is modifying an existing protocol, it would be nice to keep track of the original protocol and only trigger unsaved changes if whatever is entered is different from the original protocol. This would also allow us to later on add 'undo' or 'restore' options if helpful.

Within #169 I added a button "Revert Changes" which undoes any of the changes done in the protocol, as well as removes the "unsaved changes" notification.  While I could manually check that the latest data the user is typing matches the existing protocol data, I think it makes sense to treat any modification as an unsaved change.  Most applications in my experience treat modifications back to the original state still as "unsaved changes".  The "revert changes" can make sure the user is worry free about having edited a protocol when it mismatches with what they intended, but it turns out they did correctly adjust their protocol to the previous state. If you still think we should do the checks, I could change "trigger_unsaved_changes" to "on_field_changed" to first check if the data matches the selected protocol.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadhana-r) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:378]: I think instead of manually setting the current protocol to the loaded protocol and updating the display, at line 248 under `onDataParamterNodeModified`, when the loaded_protocols has length > 0, you can add the protocol options without the placeholder 'Select a protocol' option. Then by default, the loaded protocol will become the selected protocol and that will trigger the protocol display. Does that make sense?

[the remark this answers is not in the document]
Comment (sadhana-r) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:378]: And in the case where a user has previously selected a protocol that they are editing and then decides to Load a protocol, then the newly loaded protocol just gets added to the drop down list. It's actually better not to force the display becasue what if the user decides to load a new protocol before saving the one they were editing before. We don't want to override the display.

**Comment by sadhana-r** [comment]
> > Another thought I had, currently `trigger_unsaved_changes` is triggered anytime a user edits the textboxes. Therefore, for example, if a user changes a value from 1 to 2 and then back to 1, it still shows as having 'unsaved changes'. This is probably a separate issue, but ideally, if a user is modifying an existing protocol, it would be nice to keep track of the original protocol and only trigger unsaved changes if whatever is entered is different from the original protocol. This would also allow us to later on add 'undo' or 'restore' options if helpful.
> 
> Within #169 I added a button "Revert Changes" which undoes any of the changes done in the protocol, as well as removes the "unsaved changes" notification. While I could manually check that the latest data the user is typing matches the existing protocol data, I think it makes sense to treat any modification as an unsaved change. Most applications in my experience treat modifications back to the original state still as "unsaved changes". The "revert changes" can make sure the user is worry free about having edited a protocol when it mismatches with what they intended, but it turns out they did correctly adjust their protocol to the previous state. If you still think we should do the checks, I could change "trigger_unsaved_changes" to "on_field_changed" to first check if the data matches the selected protocol -- either works -- the overhead to this check is probably negligible.

Oh great, having the rever changes option definitely makes thinsg more worry-free. @ebrahimebrahim do you have any thoughts? If its common to treat modifications back to original state as unsaved changes, then I think it could be fine as is.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 4]
```diff
@@ -154,6 +225,187 @@ def onSceneEndClose(self, caller, event) -> None:
...
+        if len(get_openlifu_data_parameter_node().loaded_protocols) == 0:
+            placeholder_text = "No protocols to select"
+            tooltip = "Load a protocol first in order to select it for editing"
+        else:
+            placeholder_text = "Select a protocol..."
+            tooltip = ""
+
+        self.ui.protocolSelector.setToolTip(tooltip)
+        self.ui.protocolSelector.setProperty("defaultText", placeholder_text)  
+        self.ui.protocolSelector.setProperty("placeholderText", placeholder_text)
```
Comment (sadhana-r) [inline OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:254]: This is where I am suggesting that you remove the placeholder option and directly add the protocol items.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 5]
```diff
@@ -154,6 +225,187 @@ def onSceneEndClose(self, caller, event) -> None:
...
+        self.ui.protocolSelector.setCurrentText(f"{protocol.name} (ID: {protocol.id})")
+
+        self.updateWidgetSaveState(SaveState.SAVED_CHANGES)
+
+    @display_errors
+    def onDeleteProtocolFromDatabaseClicked(self, checked: bool) -> None:
+        selected_protocol = self.ui.protocolSelector.currentText
+        if selected_protocol == "Select a protocol...":
+            return
+
```
Comment (sadhana-r) [inline OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:324]: If you follow my suggesting and remove this item, all of this can be removed. Once the dropdown is enabled, a user would know to select a protocol.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:378]: Ok, and one final clarification on this point. So when any protocols are loaded at all, there should never be a time when there is a blank state where users can select protocols, where the combo box would say "Select a protocol..."; instead, at least one protocol is always selected.

[the remark this answers is not in the document]
Comment (sadhana-r) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:378]: Yep. Thats atleast how all of the other combo boxes we have put in work. We haven't been adding in placeholder text. One thought is that since a protocol is always selected/on display if loaded into Slicer, we could have the protocols be displayed but disabled/non-editable by default for the user to view the config, and add a check box or button called `Edit` which the user can select before tweaking the values to enable all the UI widgets. That way, nothing is edited by mistake.

[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:378]: Good idea.  I think I'll add the button in a separate commit to this PR.

**Comment by ebrahimebrahim** [comment]
> > > Another thought I had, currently `trigger_unsaved_changes` is triggered anytime a user edits the textboxes. Therefore, for example, if a user changes a value from 1 to 2 and then back to 1, it still shows as having 'unsaved changes'. This is probably a separate issue, but ideally, if a user is modifying an existing protocol, it would be nice to keep track of the original protocol and only trigger unsaved changes if whatever is entered is different from the original protocol. This would also allow us to later on add 'undo' or 'restore' options if helpful.
> > 
> > 
> > Within #169 I added a button "Revert Changes" which undoes any of the changes done in the protocol, as well as removes the "unsaved changes" notification. While I could manually check that the latest data the user is typing matches the existing protocol data, I think it makes sense to treat any modification as an unsaved change. Most applications in my experience treat modifications back to the original state still as "unsaved changes". The "revert changes" can make sure the user is worry free about having edited a protocol when it mismatches with what they intended, but it turns out they did correctly adjust their protocol to the previous state. If you still think we should do the checks, I could change "trigger_unsaved_changes" to "on_field_changed" to first check if the data matches the selected protocol -- either works -- the overhead to this check is probably negligible.
> 
> Oh great, having the revert changes option definitely makes things more worry-free. @ebrahimebrahim do you have any thoughts? If its common to treat modifications back to original state as unsaved changes, then I think it could be fine as is.

I agree that with the revert button coming in, the current design makes sense and is how text editors work as far as I know

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 6]
```diff
@@ -195,13 +489,64 @@ class OpenLIFUProtocolConfigLogic(ScriptedLoadableModuleLogic):
...
+        return OpenLIFUProtocolConfigParameterNode(super().getParameterNode())  # pyright: ignore[reportCallIssue]
+
+    def protocol_id_is_loaded(self, protocol_id: str) -> bool:
+        return protocol_id in get_openlifu_data_parameter_node().loaded_protocols
+
+    def protocol_id_exists(self, protocol_id: str) -> bool:
+        return self.protocol_id_is_loaded(protocol_id) or protocol_id in self.dataLogic.db.get_protocol_ids()
+
+    def database_is_loaded(self) -> bool:
+        return self.dataLogic.db is not None
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:518]: [In reference to the reply I made earlier to the part where you addressed this fix](https://github.com/OpenwaterHealth/SlicerOpenLIFU/pull/168#discussion_r1945218159), I've decided to forego my implemented solution in #169 , because editing the Data parameter node directly with this information, I think, is a more maintainable solution.

**Comment by arhowe00** [comment]
I've addressed the revisions, but I have a question before re-marking for review:

I'm currently trying to finish the edit button, but it's starting to seem like I'm going to have to cache WIP protocols for intended behavior.  To see what I mean, read the problems in the commit message of [f7ae425](https://github.com/OpenwaterHealth/SlicerOpenLIFU/pull/168/commits/f7ae425c4700e4f3663a75bf0bc820cea8d61bc3) I could either modify the `onDataParameterNodeModified()` to not `.clear()` the selectable protocols and instead monitor the `loaded_protocols` and append/remove. However, this doesn't seem like an elegant approach, as I would have to iterate over all loaded protocols and update my `protocolSelector` based on what's in there. Since both keeping new WIP protocols and WIP edits on existing protocols involves some sort of caching, I could address the unintended behaviors with protocol editing, as well as full caching of edits, in #170 .  I could also address them here if we want no qualms with a merged PR. Thoughts?

**Comment by ebrahimebrahim** [comment]
> I've addressed the revisions, but I have a question before re-marking for review:
> 
> I just added an "Edit Protocol" button, but there are some unintended behaviors. It's starting to seem like I might have to cache WIP protocols to fix things. To see more detail about the problems, see [f7ae425](https://github.com/OpenwaterHealth/SlicerOpenLIFU/pull/168/commits/f7ae425c4700e4f3663a75bf0bc820cea8d61bc3). I could either modify the `onDataParameterNodeModified()` to not `.clear()` the selectable protocols and instead monitor the `loaded_protocols` and append/remove. However, this doesn't seem like an elegant approach, as I would have to iterate over all loaded protocols and update my `protocolSelector` based on what's in there. Since both keeping new WIP protocols and WIP edits on existing protocols involves some sort of caching, I could address the unintended behaviors with protocol editing, as well as full caching of edits, in #170 . I could also address them here if we want no qualms with a merged PR. Any thoughts? Did I miss an elegant way to solve this?

I didn't yet read everything about the specific problem with the edit button, but I think I can give some general advice.

We don't want to merge a half-working thing that looks like it's supposed to work. It's okay if something is a placeholder and says that it's a placeholder.

If it's too much effort to fix the thing you added here without the caching PR, then one totally acceptable way out is to put a big banner at the top saying that this module is under construction and not everything in it works yet. That's better than wasting a lot of time to apply temporary patch fixes that will get overriden in your immediate next PR. If you do this then just document it well in this PR and the next so that it's clear what has to be done to remove the banner, and we know what to look for when reviewing this and the next PR.

**Comment by arhowe00** [comment]
So I decided to tie up loose ends with the "Edit Protocol" button, and it seemed that to keep behaviors intuitive, I had to implement some sort of caching. This is pretty close to what's needed for #170 , but it's not done in the same way. Furthermore, there is no revert changes button here, which I'll implement there. What we have should be a closed-loop without errors for editing, loading, creating new protocols, etc., with the buttons working how they should work. I'll push and request re-review.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 7]
```diff
@@ -154,6 +227,191 @@ def onSceneEndClose(self, caller, event) -> None:
...
+        if len(get_openlifu_data_parameter_node().loaded_protocols) == 0:
+            tooltip = "Load a protocol first in order to select it for editing"
+            self.ui.protocolSelector.setProperty("defaultText", "No protocols to select.")  
+            self.setProtocolEditButtonEnabled(False)
+        else:
+            tooltip = "Select a protocol..."
+            for protocol_id, protocol in get_openlifu_data_parameter_node().loaded_protocols.items():
+                text = f"{protocol.protocol.name} (ID: {protocol_id})"
+                self.ui.protocolSelector.addItem(text, protocol.protocol)
+            self.ui.protocolSelector.setCurrentIndex(prev_index)
```
Comment (sadhana-r) [inline OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:254]: When I load a session for the first time, the protocol config displays the protocol but the selector still shows 'No protocols to select' as the first option. I don't think this manually setting of the current Index is required. When the dataparameter node is modied, if new items are added, the index doesn't get changed to the new item. It stays as the prev index.

**Review by sadhana-r — CHANGES_REQUESTED** [review]
Almost there! Thanks for making all the changes. Just a few more things I found when testing the workflow that need to be fixed. Let me know if you have any questions/clarifications!

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 8]
```diff
@@ -178,8 +436,54 @@ def setParameterNode(self, inputParameterNode: Optional[OpenLIFUProtocolConfigPa
...
+            self.ui.protocolDatabaseDeleteButton.setEnabled(enabled)
+
+    def setProtocolEditButtonEnabled(self, enabled: bool) -> None:
+        self.ui.protocolEditButton.setEnabled(enabled)
+        if enabled == False:
+            self.setProtocolEditorEnabled(False)  # depends
+
+    def setAllWidgetsEnabled(self, enabled: bool) -> None:
+        self.ui.protocolSelector.setEnabled(enabled)
+        self.ui.loadProtocolFromFileButton.setEnabled(enabled)
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:479]: Since I don't yet have the functionality implemented for protocols to be loaded without a database, this button would be non-functional in this PR, and thus in this PR I assume we can only load files with a database loaded. In #171 , this functionality would be enabled. Also, I made this function so that it deliberately disables all widgets, I could make a new function that doesn't involve those two buttons, but we might want a global enable/disable for when user roles gets implemented.

[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:479]: @arhowe00 you currently _only_ have functionality for a protocol to be loaded _without_ a database right? The database-based protocol loading is coming later in #171 

I realized #171 is confusingly named -- it's about making loading work _with_ database and saving work _without_. While what currently exists is loading without and saving with.

[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:479]: Hmm, I guess we can load a protocol without a database loaded but I intended this widget only to work in this PR with a database loaded. So should I add disabling the save button without a database loaded, and enable the load protocol button?

[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:479]: > disabling the save button without a database loaded

Since that save button requires a database, yes this is probably how it should work. Disable and set a tooltip explaining why it's disabled

> Also, I made this function so that it deliberately disables all widgets, I could make a new function that doesn't involve those two buttons, but we might want a global enable/disable for when user roles gets implemented.

Good point that we will want something like this later, but let's save that for #175 because I think we'll want to have a good think about how to do it and then do it in the same way across all modules

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 7 above)
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:254]: Does this happen even if we use `.clear()`? To stay synced up with loaded protocols, I clear and reload, I assumed this would set the index to -1, add an item setting it to 0, and then load the others keeping it at 0. I did this to set it to where it was before.  I couldn't find docs on how this works, so I may not understand how the widget does this. Is what I said accurate?

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 9]
```diff
@@ -154,6 +227,191 @@ def onSceneEndClose(self, caller, event) -> None:
...
+            description = defaults["Description"],
+            pulse = openlifu_lz().bf.Pulse(frequency=defaults["Pulse frequency"], duration=defaults["Pulse duration"]),
+            focal_pattern = openlifu_lz().bf.focal_patterns.SinglePoint()
+        )
+
+        self.updateProtocolDisplayFromProtocol(protocol)
+
+        self.logic.cur_protocol_id = protocol.id
+        self.logic.cache_protocol(protocol)
+        self.logic.dataLogic.load_protocol_from_openlifu(protocol, replace_confirmed=True) # load into memory
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:305]: > One idea, is to not allow a user to 'load a protocol' until it has been saved to the database i.e. it doesn't need to show up in the data module as a loaded protocol until it has been saved to the database. This also prevents a user from performing any experiments using a protocol that hasn't been saved to the database.

I need to load the protocol because if you are editing a new protocol and then decide to also load one from the database, it updates the parameter node, triggering the selector to overwrite the WIP. For instance, if I set the index to -1 when editing a new protocol, it will set the index to a valid one when I reload the selector from loaded protocols. When editing a new protocol, I made it so the selector would reference the default fields given to the new protocol. If I temporarily set the text of the selector, it would also get washed out when updating from parameter node updates. Making this work more the most intuitive way, I believe, requires caching, which I was going to do in #170

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 7 above)
Comment (sadhana-r) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:254]: Hmm, you kight be right. I would test that out to double check. If what you're saying is right, then I would add a check that `prevIndex != -1` before setting the current index. That way theres always a valid input in the dropdown.

Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:254]: For future reference, here were the order of events:

```
# Loaded 2 protocols without selecting anything

DATA PARAMETER MODIFIED
CURRENT INDEX BEFORE CLEAR: -1
CURRENT INDEX AFTER CLEAR: -1
CURRENT INDEX AFTER ALL PROTOCOLS ADDED: 0

# Selected index 1

DATA PARAMETER MODIFIED
CURRENT INDEX BEFORE CLEAR: 1
CURRENT INDEX AFTER CLEAR: -1
CURRENT INDEX AFTER ALL PROTOCOLS ADDED: 0
```

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 10]
```diff
@@ -195,13 +499,56 @@ class OpenLIFUProtocolConfigLogic(ScriptedLoadableModuleLogic):
...
+    }
 
     def __init__(self) -> None:
         """Called when the logic class is instantiated. Can be used for initializing member variables."""
+        self.dataLogic = None
+
+        self.cur_save_state = SaveState.NO_CHANGES
+        self.cur_protocol_id = ""  # important if WIPs change the ID
+        self.cached_protocols = {}
+        self.protocols_not_in_database = set()
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:520]: Does this `_` rule also apply if the variable `cur_protocol_id` is being used by both the logic and the widget, or only if the variable is completely contained within a class?

[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:520]: The idea of the `_` rule is to communicate to the outside user (a slicer user, or someone writing some code that uses your module) that they shouldn't mess with this variable.

In your case for example with `cur_protocol_id` it seems that you want it to be gettable from the outside, but not settable (correct me if i'm wrong)

So a good way to do that in python is to put the `_` as Sadhana suggested, and then create a getter only, with no setter. Then from the outside of the class things would access it via the getter. Python makes this convenient with properties

something like this

```
class OpenLIFUProtocolConfigLogic(ScriptedLoadableModuleLogic):

   # ...

    def __init__(self, ...):
        self._cur_protocol_id = ""
        # ...

    @property
    def cur_protocol_id(self):
        return self._cur_protocol_id
```

then you can still access via `logic.cur_protocol_id` thanks to the magic of `property`

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:305]: In imminent commit, I decided to use a cache and not call `load_protocol_from_openlifu`. This behavior should be correct now.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 11]
```diff
@@ -154,6 +227,191 @@ def onSceneEndClose(self, caller, event) -> None:
...
+        self.logic.save_protocol_to_database(protocol)  # save to database
+        self.ui.protocolDatabaseDeleteButton.setEnabled(True)  # can delete now
+
+        self.logic.cached_protocols.pop(self.logic.cur_protocol_id, None)  # remove from cache
+        self.logic.protocols_not_in_database.discard(self.logic.cur_protocol_id)  # track as in database
+
+        self.saving_changes = True
+        self.logic.dataLogic.load_protocol_from_openlifu(protocol, replace_confirmed=True)  # load (if new) or reload (if changes) to memory
+        self.ui.protocolSelector.setCurrentText(f"{protocol.name} (ID: {protocol.id})")  # id might have changed
+        self.saving_changes = False
```
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:347]: This is one change I kept, and the only reason that this is needed is to maintain the notification that changes have been saved. Because when the changes are saved we wipe the cache if it was there, and in case that the user changed the ID, we also reset the official ID stored in the logic state and in the selector, `onProtocolSelectorIndexChanged` ends up getting called. We need to somehow note that just this one time (during the saving process), we don't want to set the state to NO_CHANGES. There are other ways I could deal with notifying that the current state of the module itself is saving changes, but I thought this was pretty convenient. I modified it a bit in an imminent commit so it's more obvious why I use it.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (arhowe00) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:520]: I ended up doing this for some of the things I was tracking in the widget and it made things a lot simpler to track. Great, I'll co-author both of you on that commit.

**Comment by arhowe00** [comment]
Forcepushed to rebase to main

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 12]
```diff
@@ -154,6 +245,196 @@ def onSceneEndClose(self, caller, event) -> None:
...
+        if len(get_openlifu_data_parameter_node().loaded_protocols) == 0:
+            tooltip = "Load a protocol first in order to select it for editing"
+            self.ui.protocolSelector.setProperty("defaultText", "No protocols to select.")  
+            self.setProtocolEditButtonEnabled(False)
+        else:
+            tooltip = "Select a protocol..."
+            for protocol_id, protocol_w in get_openlifu_data_parameter_node().loaded_protocols.items():
+                self.ui.protocolSelector.addItem(f"{protocol_w.protocol.name} (ID: {protocol_id})", protocol_w.protocol)
+            for protocol_id in self.logic.cached_protocols:
+                if protocol_id in self.logic.new_protocol_ids:
```
[the remark this answers is not in the document]
Comment (sadhana-r) [reply OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:271]: This loop through `new_protocol_ids` should also be outside of the `else` statement. There could be new protocols in the cache without there being loaded protocols.

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 13]
```diff
@@ -178,8 +459,70 @@ def setParameterNode(self, inputParameterNode: Optional[OpenLIFUProtocolConfigPa
...
+    def setNewProtocolWidgets(self) -> None:
+        self.setProtocolEditButtonEnabled(True)  # enable edit button (consistency)
+        self.setProtocolEditorEnabled(True)  # enable editor
+        self.ui.protocolDatabaseDeleteButton.setEnabled(False)
+
+    def setProtocolEditorEnabled(self, enabled: bool) -> None:
+        self.ui.protocolEditorSectionGroupBox.setEnabled(enabled)
+        if enabled and get_openlifu_data_parameter_node().database_is_loaded:
+            self.ui.protocolDatabaseSaveButton.setEnabled(True)
+            self.ui.protocolDatabaseDeleteButton.setEnabled(True)
```
Comment (sadhana-r) [inline OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:498]:
```suggestion
            self.setDatabaseButtonsEnabled(True)
```

File: OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py [hunk 14]
```diff
@@ -178,8 +459,70 @@ def setParameterNode(self, inputParameterNode: Optional[OpenLIFUProtocolConfigPa
...
+        elif not enabled:
+            self.ui.protocolDatabaseSaveButton.setEnabled(False)
+            self.ui.protocolDatabaseDeleteButton.setEnabled(False)
```
Comment (sadhana-r) [inline OpenLIFUProtocolConfig/OpenLIFUProtocolConfig.py:501]:
```suggestion
            self.SetDatabaseButtonsEnabled(False)
```

**Review by sadhana-r — CHANGES_REQUESTED** [review]
Great work! Last round of very minor suggestions/fixes before this gets merged. 

There are also a couple commit messages missing the reference to (#164). You can do an interactive rebase and use `reword` to edit those.

**Review by sadhana-r — COMMENTED** [review]
I think we need to restructure how the buttons/widgets are being enabled and diabled. There are too many different functions enabling/disabling the same buttons that are overriding each other. For example, if I open the protocol config module, and connect a database, the `Create New Protocol` should be enabled. In `onDataParamaterNodeModified`, the check for `database_is_loaded` should enable the button, but a few lines later, when checking for loaded protocols `SetProtocolEditButtonEnabled` is set to False, which sets the `EditorEnabled` to False, which sets `SetDatabaseButtonsEnabled` to False. 

I think we can approve and merge this PR since nothing is broken, and we can open a separate issue to streamline the different enable/disable conditions in a different PR. @ebrahimebrahim @arhowe00 what do you think?

**Comment by arhowe00** [comment]
> I think it may be a simple fix in that the `CreateNewProtocol` button should be controlled independent of the Save/Delete buttons. @arhowe00 let me know if you prefer to implement the fix in a separate PR.

Ahh yes, now I'm seeing why I had this weird thing which you noted below where I was toggling just those buttons.  I can't tell if I didn't have the Create New Protocol button a part of that, but I agree that it makes more sense to have the Create New Protocol button activated separately. 

image

**Comment by sadhana-r** [comment]
I'm going to go ahead and approve and merge this PR. I still think there are aspects of the button enabling/disabling that need to be re-organized. For example, when I create a new protocol (without a database) but then load a database after, the delete from database button is acive even though the new protocol isn't in the database. It will be easier to test out these different edge cases once the manual workflow (without database connection) has been implemented in #171.

**Review by sadhana-r — APPROVED** [review]
